diff --git a/dist/index.js b/dist/index.js
index b1f84226292a45d5e2df6e17fb16528b0d0df641..40cd9673090792c25914b267cf7d2ebca5503788 100644
--- a/dist/index.js
+++ b/dist/index.js
@@ -84,9 +84,7 @@ function createVentoInterface(options) {
     if (env.utils._11tyCtx?.page?.inputPath === newContext?.page?.inputPath) {
       return;
     }
-    for (const K of CONTEXT_DATA_KEYS) {
-      env.utils._11tyCtx[K] = newContext[K];
-    }
+    env.utils._11tyCtx = newContext;
     DEBUG.setup("Reload context, new context is: %o", env.utils._11tyCtx);
   }
   function loadPlugins(plugins) {
